reactor-core/src/main/java/reactor/core/publisher/FluxBufferPredicate.java

```diff
@@ -205,25 +213,31 @@ public boolean tryOnNext(T t) {
 				return true;
 			}
 
-			boolean requestMore;
 			if (mode == Mode.UNTIL && match) {
 				b.add(t);
-				requestMore = onNextNewBuffer();
+				onNextNewBuffer();
 			}
 			else if (mode == Mode.UNTIL_CUT_BEFORE && match) {
-				requestMore = onNextNewBuffer();
+				onNextNewBuffer();
 				b = buffer;
 				b.add(t);
 			}
 			else if (mode == Mode.WHILE && !match) {
-				requestMore = onNextNewBuffer();
+				onNextNewBuffer();
 			}
 			else {
 				b.add(t);
-				return !(!fastpath && requested != 0);
 			}
 
-			return !requestMore;
+			boolean isNotExpectingFromSource = REQUESTED_FROM_SOURCE.decrementAndGet(this) == 0;
```

Comment:
it will take a lot of time to go down to zero, but anyway why to decrement for a fast path?

Reply:
not sure what you mean. maybe that this decrement operation should be skipped in case `fastpath == true` ?

Reply:
we need to keep separate track of "how many buffers are wanted by downstream" (`REQUESTED_BUFFERS`) vs "how many elements have been requested upstream" (`REQUESTED_FROM_SOURCE`), so that when we're filling a buffer and we see that said buffer is not yet full, we only ask the source for one more element IFF there is no pending request to it already. Otherwise, we know that we should soon receive at least another element which might fulfill the buffer.

Reply:
> maybe that this decrement operation should be skipped in case fastpath == true ?

yes, I mean this

Reply:
updated, by extracting `if (fastpath) return true;` we take care of the fastpath case more clearly and avoid these atomic operations altogether